[Shortcut Guide] Support delayed display of shortcuts except for taskbar shortcuts

OkamiWong opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

Support delayed display of shortcuts except for taskbar shortcuts as described in #21403.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [ ] **Closes:** #21403 
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [x] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here:
   - https://github.com/MicrosoftDocs/windows-dev-docs/pull/4151

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

Currently the users can only activate all the shortcut guides at once. However, some users requested a way to activate the taskbar hints only (Related issues: #6265 #8765 #11931).

My solution is to allow the users to set different press time before showing the "global windows shortcuts" and the "taskbar icon shortctus".

By setting the press time before showing the taskbar icon shortcuts smaller than the global windows shortcuts, the screens of the users who only want to check the Win + [Number] shortcuts are not distrupted by the rest shortcut guides anymore. The user can also chose to show global windows shortcuts ahead of the taskbar icon shortcuts as well.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

- If the user sets the press time before showing the taskbar icon shortcuts shorter than the global windows shortcuts, the taskbar icon shortcuts will be shown ahead of the global windows shortcuts, and vice versa.
  - ![GIF 2022-11-12 20-45-10](https://user-images.githubusercontent.com/24769875/201474538-24e0d14d-88df-4deb-811d-f197a09a1a87.gif)
  - ![GIF 2022-11-12 20-43-14](https://user-images.githubusercontent.com/24769875/201474540-131b9b66-80d3-4515-8491-38cd98eb615d.gif)
- If the user sets the same press time before showing the taskbar icon shortcuts and the global windows shortcuts, the behavior is the same as before.
  - ![GIF 2022-11-12 20-41-28](https://user-images.githubusercontent.com/24769875/201474529-ef9fbad5-8376-4465-83c8-cf7e01f95a65.gif)
- If the shortcut guide is activated by customized shortcut, the behavior is also the same as before.
  - ![GIF 2022-11-12 21-03-24](https://user-images.githubusercontent.com/24769875/201475344-bcd1d75c-efbf-40ee-a64e-5b84b995d63b.gif)

OkamiWong commented:
@microsoft-github-policy-service agree

htcfreek commented:
Great feature. What about making the background fully transparent in case of delayed showing (when only the taskbar numbers are shown)?

OkamiWong commented:
> Great feature. What about making the background fully transparent in case of delayed showing (when only the taskbar numbers are shown)?

It is feasible to do that. However, I think the semi-transparent background can make the taskbar numbers more readable and lead the focus of the user to the position of the taskbar numbers.

Jay-o-Way commented:
Interesting. Thanks for picking this up.

Wanna ask to align the text of both settings (Press duration) so they look alike, and to make it more clear what setting is for what area. Actually, if the .header is good, we shouldn't need the .descriptions. For example:
| .header | |
|:---|---:|
| **Press duration for global Windows shortcuts (ms)** | [&nbsp;&nbsp;&nbsp;500] |
| **Press duration for taskbar icon shortcuts (ms)** | [ 2000] |

I notice a difference in animation. In the "original" version there's a short "slide+tranparency", is it possible to apply the same to all sections?

Jay-o-Way commented on src/modules/ShortcutGuide/ShortcutGuide/ShortcutGuideSettings.h:
```diff
@@ -9,4 +9,5 @@ struct ShortcutGuideSettings
     std::wstring disabledApps = L"";
     bool shouldReactToPressedWinKey = false;
     int windowsKeyPressTime = 900;
+    int delayAfterShowingTaskbarShortcuts = 900;
```
This variable name could be better. Please look at both and align.

Jay-o-Way commented on src/modules/ShortcutGuide/ShortcutGuide/overlay_window.cpp:
```diff
@@ -422,6 +422,11 @@ void D2DOverlayWindow::apply_overlay_opacity(float opacity)
     overlay_opacity = opacity;
 }
 
+void D2DOverlayWindow::apply_delay_after_showing_taskbar_shortcuts(int delay)
```
Same question for the name: "delay _after_ showing" is incorrect

Jay-o-Way commented on src/modules/ShortcutGuide/ShortcutGuide/overlay_window.h:
```diff
@@ -51,6 +51,7 @@ class D2DOverlayWindow : public D2DWindow
     void show(HWND active_window, bool snappable);
     ~D2DOverlayWindow();
     void apply_overlay_opacity(float opacity);
+    void apply_delay_after_showing_taskbar_shortcuts(int delay);
```
Same here

Jay-o-Way reviewed (changes requested):
I'm thinking of two similar and independant parts now. What I would call "global shortcuts" and "taskbar icon shortcuts"

OkamiWong commented:
Thanks for your great suggestions. @Jay-o-Way 

> *Actually, if the .header is good, we shouldn't need the .descriptions.*

Will remove the .descriptions of the two settings.

> *I notice a difference in animation. In the "original" version there's a short "slide+tranparency", is it possible to apply the same to all sections?*

Yeah, there should be animation when displaying the "global shortcuts".

> *I'm thinking of two similar and independant parts now. What I would call "global shortcuts" and "taskbar icon shortcuts"*

I like this idea. We should do it.

I will update the settings descriptions, apply the animation, and also change the naming in the next couple of days.

OkamiWong commented:
@Jay-o-Way Hi, could you please review again? I just pushed the improvements and updated the description of the PR. I also updated the demo gifs as well, please take a look.

Jay-o-Way commented:
At quick glance looks good. I'll look again when I'm back on comp.

Jay-o-Way commented:
Hm, can't seem to _activate_ the Shortcut Guide from this pr on my computer. I must be missing something..?

crutkas commented:
clever.  This feature took me a second to actually parse what it was doing!

crutkas commented:
@OkamiWong please fix the merge conflict and then i'd like to get a review done so we can merge this in.  neat.

crutkas commented:
when you are ready, click the "ready for review" button so it isn't a draft anymore

Jay-o-Way commented:
Works great!
Just one thing - but I'm not sure if it's related to this pr at all... ScG only appears on my primary monitor, not the secondary.

OkamiWong commented:
> Works great! Just one thing - but I'm not sure if it's related to this pr at all... ScG only appears on my primary monitor, not the secondary.

Thanks for reviewing! Just confirmed that this is not related to this PR by tried the ScG on the latest main branch.

OkamiWong commented:
Hi @Jay-o-Way . When will this get merged? I noticed that you approved this PR but it still says the merging is blocked and review is required.

htcfreek commented:
@OkamiWong 
First we need an approval from a core team member like @crutkas .

jaimecbernardo commented on src/modules/ShortcutGuide/ShortcutGuideModuleInterface/dllmain.cpp:
```diff
@@ -185,9 +185,11 @@ class ShortcutGuideModule : public PowertoyModuleIface
     // If the module should be activated through the legacy pressing windows key behavior.
-    const UINT DEFAULT_MILLISECONDS_WIN_KEY_SHOULD_BE_PRESSED = 900;
+    const UINT DEFAULT_MILLISECONDS_WIN_KEY_PRESS_TIME_FOR_GLOBAL_WINDOWS_SHORTCUTS = 1000;
+    const UINT DEFAULT_MILLISECONDS_WIN_KEY_PRESS_TIME_FOR_TASKBAR_ICON_SHORTCUTS = 1000;
```
Why did we change the default to 1000 from 900?

OkamiWong replied:
We can keep 900. I just thought that 1000 is a more natural number. Will change it back to 900.

jaimecbernardo reviewed (changes requested):
Hi, Thanks a lot for the contribution! Nice feature!
Got some questions, but other than that it looks good! Great work!

jaimecbernardo commented on src/settings-ui/Settings.UI.Library/ShortcutGuideProperties.cs:
```diff
@@ -12,7 +12,8 @@ public ShortcutGuideProperties()
             UseLegacyPressWinKeyBehavior = new BoolProperty(false);
-            PressTime = new IntProperty(900);
+            PressTimeForGlobalWindowsShortcuts = new IntProperty(1000);
+            PressTimeForTaskbarIconShortcuts = new IntProperty(1000);
```
Missed this one ;)

OkamiWong replied:
Thanks you for your swift response!

jaimecbernardo reviewed (approved):
LGTM! Missed one, but I can push that. I'll merge it after CI runs.
Thanks a lot for the contribution!
